src/consumption.cpp

```diff
@@ -732,6 +732,19 @@ void player::consume_effects( item &food, bool rotten )
             fun = 1;
         }
     }
+    //Food is made out of meat (use ALLERGEN_MEAT, not CARNIVORE_OK (the latter has eggs and milk),
+    //Reduces bad morale penalties on food only made of meat.
+    if( has_trait( "CARNIVORE" ) && food.has_flag( "ALLERGEN_MEAT" ) &&
+        !food.has_any_flag( carnivore_blacklist ) &&
+        !( food.has_flag( "CANNIBALISM" ) && !has_trait( "CANNIBAL" ) ) ) {
```

Comment:
What about sapiovore and psycho? You can check them all by `has_trait_flag( "CANNIBAL" )` - this is a relatively new function to make consumption checks more clear. It checks all the cannibal type traitsat once.